cli/src/commands/metadata.rs

```diff
@@ -7,37 +7,38 @@ use color_eyre::eyre;
 use frame_metadata::{RuntimeMetadata, RuntimeMetadataPrefixed};
 use jsonrpsee::client_transport::ws::Uri;
 use scale::{Decode, Encode};
-use std::io::{self, Write};
-use subxt_codegen::utils::fetch_metadata_hex;
+use std::{
+    io::{self, Write},
+    path::PathBuf,
+};
 use subxt_metadata::retain_metadata_pallets;
 
+use super::utils::MetadataSource;
+
 /// Download metadata from a substrate node, for use with `subxt` codegen.
 #[derive(Debug, ClapParser)]
 pub struct Opts {
-    /// The url of the substrate node to query for metadata.
-    #[clap(
-        name = "url",
-        long,
-        value_parser,
-        default_value = "http://localhost:9933"
-    )]
-    url: Uri,
+    /// The url of the substrate node to query for metadata for codegen.
+    #[clap(name = "url", long, value_parser)]
+    url: Option<Uri>,
+    /// The path to the encoded metadata file.
+    #[clap(short, long, value_parser)]
+    file: Option<PathBuf>,
     /// The format of the metadata to display: `json`, `hex` or `bytes`.
     #[clap(long, short, default_value = "bytes")]
     format: String,
     /// Generate a subset of the metadata that contains only the
     /// types needed to represent the provided pallets.
     #[clap(long, use_value_delimiter = true, value_parser)]
-    retain_pallets: Option<Vec<String>>,
+    pallets: Option<Vec<String>>,
 }
 
 pub async fn run(opts: Opts) -> color_eyre::Result<()> {
-    let hex_data = fetch_metadata_hex(&opts.url).await?;
-
-    let bytes = hex::decode(hex_data.trim_start_matches("0x"))?;
+    let source = MetadataSource::new(opts.url, opts.file)?;
```

Comment:
Sorry; I think I wasn't very clear; this `MetadataSource` thing could be the value of one of the Opts directly (perhaps behind a `--source` flag), rather than being some glorified way to call a function to use either file or URL :)

An alternate approach that might be nice would be a struct like:

```rust
#[derive(Args)]
struct FileOrUrl {
    #[clap(...)]
    pub file: Option<PathBuf>
    #[clap(...)]
    pub url: Option<Uri>
}
```

And then using that with `#[command(flatten)]` in the opts, something like:

```rust
pub struct Opts {
    ...
    #[command(flatten)]
    file_or_url: FileOrUrl,
```

(with whatever annotation prevents both being used together if that's possible). Search for `flatten` in this for an example: https://docs.rs/clap/latest/clap/_derive/_tutorial/index.html.

The `enum MetadataSource` is more "correct" in that it allows only one of a URL or file path, but the above allows for either `--file` or `--url` to be used. Maybe theres an even better way, but yeah I just thought we could standardise how we accept the metadata location in some way :)